### PR TITLE
Autoindent improvements

### PR DESCRIPTION
**Pull request by harshad1**
This PR addresses #882 and #336. #869 is partially addressed as well

Fixes / changes include:

1. Change unordered maximum leading spaces to 16
2. Handle case where last inserted character is a newline (this happens when android fixes a spelling)
3. Simplify handling of leading space in auto indent. Behavior is now like a text editor auto indent. The leading indent of the current line is always carried over. This automatically addresses indent in indented code block.
4. In a list, if <CR> (enter) is pressed in an empty list, that element is removed. (See https://github.com/gsantner/markor/issues/336#issuecomment-605091437) 

All of these changes only affect the markdown format as I found making all the necessary changes too hard to do.

### CODE REVIEW

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java [hunk 1]
```diff
@@ -13,17 +13,14 @@
...
     public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
         try {
-            if (end - start == 1
-                    && start < source.length()
-                    && dstart <= dest.length()) {
-                char newChar = source.charAt(start);
-
-                if (newChar == '\n') {
+            if (start < source.length() && dstart <= dest.length()) {
+                if (source.charAt(start) == '\n' || source.charAt(end - 1) == '\n') {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java]: Codacy Issue found: [These nested if statements could be combined](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownHighlighterPattern.java [hunk 2]
```diff
@@ -17,7 +17,7 @@
     HEADING(Pattern.compile("(?m)((^#{1,6}[^\\S\\n][^\\n]+)|((\\n|^)[^\\s]+.*?\\n(-{2,}|={2,})[^\\S\\n]*$))")),
     HEADING_SIMPLE(Pattern.compile("(?m)(#{1,6}\\s.*$)")),
     LINK(Pattern.compile("\\[([^\\[]+)\\]\\(([^\\)]+)\\)")),
-    LIST_UNORDERED(Pattern.compile("(\\n|^)\\s{0,12}([*+-])( \\[[ xX]\\])?(?= )")),
+    LIST_UNORDERED(Pattern.compile("(\\n|^)\\s{1,16}([*+-])( \\[[ xX]\\])?(?= )")),
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/markdown/MarkdownHighlighterPattern.java]: I made this change to handle 4 indent levels. I can revert this change if it is likely to cause issues elsewhere

Comment (gsantner) [reply app/src/main/java/net/gsantner/markor/format/markdown/MarkdownHighlighterPattern.java]: tbh, if you make 4 indent levels in markdown you make something wrong :D. Just whitespace a whole line then a list character and horizontal space is over :D

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java [hunk 3]
```diff
@@ -53,49 +50,34 @@ private int findLineBreakPosition(Spanned dest, int dstart) {
...
+            indentString += uMatch.group() + " ";
         } else {
-            return "";
+            Matcher oMatch = MarkdownHighlighterPattern.LIST_ORDERED.pattern.matcher(dest.toString().substring(iend, dend));
+            if (oMatch.find()) {
+                indentString += addNumericListItemIfNeeded(oMatch.group(1));
+            }
         }
+
+        return indentString;
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java]: This simplified structure handles all cases including first line, last line and selections spanning lines.

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java [hunk 4]
```diff
@@ -13,17 +13,14 @@
...
     public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
         try {
-            if (end - start == 1
-                    && start < source.length()
-                    && dstart <= dest.length()) {
-                char newChar = source.charAt(start);
-
-                if (newChar == '\n') {
+            if (start < source.length() && dstart <= dest.length()) {
+                if ((source.charAt(start) == '\n') || (source.charAt(end - 1) == '\n')) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java]: Codacy Issue found: [These nested if statements could be combined](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java [hunk 5]
```diff
@@ -13,17 +13,14 @@
...
     public CharSequence filter(CharSequence source, int start, int end, Spanned dest, int dstart, int dend) {
         try {
-            if (end - start == 1
-                    && start < source.length()
-                    && dstart <= dest.length()) {
-                char newChar = source.charAt(start);
-
-                if (newChar == '\n') {
+            if (start < source.length() && dstart <= dest.length()) {
+                if (isNewLine(source, start, end)) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java]: Codacy Issue found: [These nested if statements could be combined](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java (hunk 5 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java]: Codacy Issue found: [These nested if statements could be combined](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownHighlighterPattern.java (hunk 2 above)
Comment (harshad1) [reply app/src/main/java/net/gsantner/markor/format/markdown/MarkdownHighlighterPattern.java]: True :). However I write a lot on the desktop and make small changes on the go in Markor. On the desktop I use lots of indentation and so I would like to support it in Markor as well. 

Is there a reason we can't use \\s*?

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

**Comment by harshad1** [comment]
If we decide on an approach, I can fix https://github.com/gsantner/markor/issues/336 as a part of this PR.

**Comment by gsantner** [comment]
Please implement what you think fits best.

Please take one additional thing in mind: When having new todo item from today, like 

`2021-03-27^` 

and pos is at ^, pressing enter removes the todo line

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java [hunk 6]
```diff
@@ -11,28 +11,31 @@
 
 import android.text.InputFilter;
 import android.text.Spanned;
+import android.util.Log;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java]: Codacy Issue found: [Unused import - android.util.Log.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 7]
```diff
@@ -102,6 +107,57 @@ public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                 }
             }
         });
+
+        addTextChangedListener(new TextWatcher() {
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+                return;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Avoid unnecessary return statements](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java (hunk 6 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java]: Codacy Issue found: [Avoid unused imports such as 'android.util.Log'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 8]
```diff
@@ -102,6 +107,57 @@ public void beforeTextChanged(CharSequence s, int start, int count, int after) {
...
+            }
+
+            @Override
+            public void onTextChanged(CharSequence s, int start, int before, int count) {
+
+                // This method is called to notify you that, within s,
+                // the count characters beginning at start have just replaced old text that had length before.
+                if ( count > 0 && start < s.length() && s.charAt(start) == '\n' ) {
+                    // Check if previous like matches
+                    int iStart, iEnd;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Use one line for each declaration, it enhances code readability.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 9]
```diff
@@ -102,6 +107,55 @@ public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                 }
             }
         });
+
+        addTextChangedListener(new TextWatcher() {
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
+                return;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Avoid unnecessary return statements](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 10]
```diff
@@ -102,6 +107,53 @@ public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                 }
             }
         });
+
+        addTextChangedListener(new TextWatcher() {
+            @Override
+            public void beforeTextChanged(CharSequence s, int start, int count, int after) { }
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Document empty method body](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

**Comment by gsantner** [comment]
probably fixes #869 as well?

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 11]
```diff
@@ -14,16 +14,20 @@
 import android.support.v7.widget.AppCompatEditText;
 import android.text.Editable;
 import android.text.InputFilter;
+import android.text.Spannable;
+import android.text.Spanned;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Avoid unused imports such as 'android.text.Spanned'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 11
           

Complexity increasing per file
==============================
- app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java  10
- app/src/main/java/net/gsantner/markor/util/StringUtils.java  4
         

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

File: app/src/main/java/net/gsantner/markor/util/StringUtils.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        return i;
+    }
+
+    public static int getNextNonWhitespace(CharSequence s, int start) {
+        return getNextNonWhitespace(s, start, s.length());
+    }
+
+    public static int getNextNonWhitespace(CharSequence s, int start, int maxRange) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/util/StringUtils.java]: Codacy Issue found: [Avoid reassigning parameters such as 'maxRange'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java (hunk 11 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Unused import - android.text.Spanned.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 13]
```diff
@@ -14,16 +14,20 @@
...
 import android.text.TextWatcher;
 import android.util.AttributeSet;
 import android.view.KeyEvent;
 
 import net.gsantner.markor.activity.MainActivity;
+import net.gsantner.markor.format.markdown.MarkdownHighlighterPattern;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Avoid unused imports such as 'net.gsantner.markor.format.markdown.MarkdownHighlighterPattern'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/util/StringUtils.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        return i;
+    }
+
+    public static int getLineEnd(CharSequence s, int start) {
+        return getLineEnd(s, start, s.length());
+    }
+
+    public static int getLineEnd(CharSequence s, int start, int maxRange) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/util/StringUtils.java]: Codacy Issue found: [Avoid reassigning parameters such as 'maxRange'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 15]
```diff
@@ -14,16 +14,20 @@
...
 import net.gsantner.markor.model.Document;
 import net.gsantner.markor.util.AppSettings;
 import net.gsantner.markor.util.ContextUtils;
 
 import java.io.File;
+import java.util.regex.Matcher;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Unused import - java.util.regex.Matcher.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.regex.Matcher'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java [hunk 16]
```diff
@@ -14,16 +14,20 @@
 import android.support.v7.widget.AppCompatEditText;
 import android.text.Editable;
 import android.text.InputFilter;
+import android.text.Spannable;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Avoid unused imports such as 'android.text.Spannable'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java (hunk 13 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Unused import - net.gsantner.markor.format.markdown.MarkdownHighlighterPattern.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java (hunk 16 above)
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/HighlightingEditor.java]: Codacy Issue found: [Unused import - android.text.Spannable.](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

File: app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+import net.gsantner.markor.format.markdown.MarkdownHighlighterPattern;
+import net.gsantner.markor.util.StringUtils;
+
+import java.util.regex.Matcher;
+
+public class ListHandler implements TextWatcher {
+
+    @Override
+    public void onTextChanged(CharSequence s, int start, int before, int count) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java]: Codacy Issue found: [Avoid reassigning parameters such as 'start'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java  9
- app/src/main/java/net/gsantner/markor/util/StringUtils.java  4
         

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

File: app/src/main/java/net/gsantner/markor/util/StringUtils.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        return i;
+    }
+
+    public static int getNextNonWhitespace(CharSequence s, int start) {
+        return getNextNonWhitespace(s, start, s.length());
+    }
+
+    public static int getNextNonWhitespace(CharSequence s, int start, int maxRange) {
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/util/StringUtils.java]: Codacy Issue found: [Avoid reassigning parameters such as 'maxRange'](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231)

**Comment by harshad1** [comment]
@gsantner This PR now includes a number of cleanups and improvements. I need to test it more before it is ready to merge. If possible, could you test it as well? Especially edge cases near the start and end of the file. Performance on very large files should probably be checked as well.

File: app/src/main/java/net/gsantner/markor/format/markdown/MarkdownHighlighterPattern.java [hunk 19]
```diff
@@ -17,8 +17,8 @@
     HEADING(Pattern.compile("(?m)((^#{1,6}[^\\S\\n][^\\n]+)|((\\n|^)[^\\s]+.*?\\n(-{2,}|={2,})[^\\S\\n]*$))")),
     HEADING_SIMPLE(Pattern.compile("(?m)(#{1,6}\\s.*$)")),
     LINK(Pattern.compile("\\[([^\\[]+)\\]\\(([^\\)]+)\\)")),
-    LIST_UNORDERED(Pattern.compile("(\\n|^)\\s{0,12}([*+-])( \\[[ xX]\\])?(?= )")),
-    LIST_ORDERED(Pattern.compile("(?m)^([0-9]+)(\\.)")),
+    LIST_UNORDERED(Pattern.compile("(\\n|^)\\s{0,16}([*+-])( \\[[ xX]\\])?(?= )")),
+    LIST_ORDERED(Pattern.compile("(?m)^([0-9]+)(\\. )")),
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/markdown/MarkdownHighlighterPattern.java]: trailing space for ordered list number

File: app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+
+        if ( count > 0 && start > -1 && start < s.length() && s.charAt(start) == '\n' ) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java]: Detects if enter pressed on empty list (correctly handles indent) and marks line for deletion.

File: app/src/main/java/net/gsantner/markor/util/StringUtils.java [hunk 21]
```diff
@@ -0,0 +1,54 @@
+package net.gsantner.markor.util;
+
+public final class StringUtils {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/util/StringUtils.java]: We seem to be writing these routines over and over again. Moved them all here.

This module ought to be tested

**Comment by harshad1** [comment]
@gsantner This PR is ready to merge unless you have additional comments / issues.

File: app/src/main/java/net/gsantner/markor/util/StringUtils.java (hunk 21 above)
Comment (gsantner) [reply app/src/main/java/net/gsantner/markor/util/StringUtils.java]: Can you move this to the opoc/util package?

File: app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java (hunk 20 above)
Comment (gsantner) [reply app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java]: You can add exactly this comment as comment in code above the line

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java  9
- app/src/main/java/net/gsantner/markor/util/StringUtils.java  5
         

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java  -1
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/src/main/java/net/gsantner/opoc/util/StringUtils.java  5
- app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java  9
         

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java  -1
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

File: app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java [hunk 22]
```diff
@@ -204,6 +215,57 @@ public void onClick(DialogInterface dialog, int selectedColor, Integer[] allColo
...
+            else {
+                text.insert(lineStart, tabString);
+                selectionEnd += _tabWidth;
+            }
+
+            text = _hlEditor.getText();
+            // Get next line
+            lineStart = StringUtils.getLineEnd(text, lineStart, selectionEnd) + 1;
+        }
+    }
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java:253]: Moved `runIndentLines` routine from TextActions to CommonTextActions and hooked it up to the indent and de-indent keyboard shortcut.

File: app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java [hunk 23]
```diff
@@ -55,9 +61,16 @@
...
+    protected AppSettings _appSettings;
+    private int _tabWidth;
+    protected Context _context;
+
     public CommonTextActions(Activity activity, HighlightingEditor hlEditor) {
         _activity = activity;
         _hlEditor = hlEditor;
+        _context = activity != null ? activity : _hlEditor.getContext();
+        _appSettings = new AppSettings(_context);
+        _tabWidth = _appSettings.getTabWidth();
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java:73]: I am deriving _tabWidth from settings, while the existing indent function uses `_hlEditor.getShiftWidth`. `getShiftWidth` appears to search for strings such as `"sw=2"` in the text and will default to 4 if these are not found. I am assuming using the app settings value is the correct thing to do here.

File: app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java [hunk 24]
```diff
@@ -108,11 +121,9 @@ public boolean runAction(String action) {
                     } else if (callbackPayload.equals(rstr(R.string.char_punctation_mark_arrows))) {
                         _hlEditor.insertOrReplaceTextOnCursor("»«");
                     } else if (callbackPayload.equals(rstr(R.string.indent))) {
-                        _hlEditor.simulateKeyPress(KeyEvent.KEYCODE_MOVE_HOME);
-                        indentCurrentLine();
+                        runIndentLines(false);
                     } else if (callbackPayload.equals(rstr(R.string.deindent))) {
-                        _hlEditor.simulateKeyPress(KeyEvent.KEYCODE_MOVE_HOME);
-                        deIndentCurrentLine();
+                        runIndentLines(true);
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java:126]: I have not removed `indentCurrentLine` `deIndentCurrentLine`. It appears they are not used anywhere. @gsantner Would you like me to delete them?

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- app/src/main/java/net/gsantner/opoc/util/StringUtils.java  5
- app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java  9
         

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -6
+ app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java  9
- app/src/main/java/net/gsantner/opoc/util/StringUtils.java  5
         

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java  -1
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

**Comment by harshad1** [comment]
@gsantner please let me know if you'd like any more changes before you merge.

File: app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java (hunk 23 above)
Comment (gsantner) [reply app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java:73]: ok

File: app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java (hunk 24 above)
Comment (gsantner) [reply app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java:126]: you can remove old functions

**Comment by gsantner** [comment]
@harshad1 
Thanks, yes you can remove the old two functions. Also please do a Auto-Code Reformat in Android studio for whole java code (i.e. usually strf+shift+f on  "net" package)

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/src/main/java/net/gsantner/opoc/util/StringUtils.java  5
- app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java  9
         

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -6
+ app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

**Comment by harshad1** [comment]
@gsantner removed unneeded functions and ran code format

**Comment by gsantner** [comment]
ok thanks!

does this also fix #891?

File: app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java [hunk 25]
```diff
@@ -55,9 +61,16 @@
...
     private final HighlightingEditor _hlEditor;
 
+    protected AppSettings _appSettings;
+    private int _tabWidth;
+    protected Context _context;
+
     public CommonTextActions(Activity activity, HighlightingEditor hlEditor) {
         _activity = activity;
         _hlEditor = hlEditor;
+        _context = activity != null ? activity : _hlEditor.getContext();
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java:71]: I cant find somethinz requiring saving context as member. This can be removed right? Notice: It may leak memory/data

File: app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java [hunk 26]
```diff
@@ -55,9 +61,16 @@
     private final Activity _activity;
     private final HighlightingEditor _hlEditor;
 
+    protected AppSettings _appSettings;
+    private int _tabWidth;
+    protected Context _context;
```
Comment (gsantner) [inline app/src/main/java/net/gsantner/markor/format/general/CommonTextActions.java:66]: We dont need to preserve context as member right? Please remove as it may leak memory/data

**Comment by gsantner** [comment]
@harshad1
Thanks for your efford. You know now better than me and everybody else how this all works together :D.
I'll merge this on computer then. :tada:

**Comment by gsantner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/src/main/java/net/gsantner/markor/ui/hleditor/ListHandler.java  9
- app/src/main/java/net/gsantner/opoc/util/StringUtils.java  5
         

Complexity decreasing per file
==============================
+ app/src/main/java/net/gsantner/markor/format/markdown/MarkdownAutoFormat.java  -6
+ app/src/main/java/net/gsantner/markor/ui/hleditor/TextActions.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/gsantner/markor/pullRequest?prid=5254231&bid=17216451)

**Comment by harshad1** [comment]
> @harshad1
> Thanks for your efford. You know now better than me and everybody else how this all works together :D.
> I'll merge this on computer then. tada

Feedback is always appreciated! I am actually very new to android development.
